backend/session/libseat.c

```diff
@@ -0,0 +1,184 @@
+	wl_list_init(&session->devices);
+
+	session->seat = libseat_open_seat(&seat_listener, session);
+	if (session->seat == NULL) {
+		wlr_log_errno(WLR_ERROR, "Unable to create seat");
+		goto error;
+	}
+
+	const char *seat_name = libseat_seat_name(session->seat);
+	if (seat_name == NULL) {
+		wlr_log_errno(WLR_ERROR, "Unable to get seat info");
+		goto error;
+	}
+	snprintf(session->base.seat, sizeof(session->base.seat), "%s", seat_name);
+
+	struct wl_event_loop *event_loop = wl_display_get_event_loop(disp);
+	session->event = wl_event_loop_add_fd(event_loop, libseat_get_fd(session->seat),
+		WL_EVENT_READABLE, libseat_event, session->seat);
+
+	// We may have received enable_seat immediately after the open_seat result,
+	// so, dispatch once without timeout to speed up activation.
+	if (libseat_dispatch(session->seat, 0) == -1) {
+		wlr_log_errno(WLR_ERROR, "libseat dispatch failed");
+		goto error;
+	}
+
+	wlr_log(WLR_INFO, "Successfully loaded libseat session");
+	session->base.impl = &session_libseat;
+	return &session->base;
+
+error:
+	if (session->seat != NULL) {
+		libseat_close_seat(session->seat);
+		session->seat = NULL;
+	}
+	if (session->event != NULL) {
+		wl_event_source_remove(session->event);
+		session->event = NULL;
+	}
+	free(session);
+	return NULL;
+}
+
+static void libseat_session_destroy(struct wlr_session *base) {
+	struct libseat_session *session = libseat_session_from_session(base);
+
+	libseat_close_seat(session->seat);
+	wl_event_source_remove(session->event);
+	free(session);
+}
+
+static struct libseat_device *find_device_by_fd(struct libseat_session *session, int fd) {
+	struct libseat_device *dev;
+	wl_list_for_each(dev, &session->devices, link) {
+		if (dev->fd == fd) {
+			return dev;
+		}
+	}
+	return NULL;
+}
+
+static int libseat_session_open_device(struct wlr_session *base, const char *path) {
+	struct libseat_session *session = libseat_session_from_session(base);
+
+	int fd;
+	int device_id = libseat_open_device(session->seat, path, &fd);
+	if (device_id == -1) {
+		wlr_log_errno(WLR_ERROR, "Failed to open device '%s'", path);
+		return -1;
+	}
+
+	struct libseat_device *dev = calloc(1, sizeof(struct libseat_device));
+	if (dev == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		libseat_close_device(session->seat, device_id);
+		return -1;
+	}
+
+	dev->fd = fd;
+	dev->device_id = device_id;
+	wl_list_insert(&session->devices, &dev->link);
+
+	return fd;
+}
+
+static void libseat_session_close_device(struct wlr_session *base, int fd) {
+	struct libseat_session *session = libseat_session_from_session(base);
+
+	struct libseat_device *dev = find_device_by_fd(session, fd);
+	if (dev == NULL) {
+		wlr_log(WLR_ERROR, "No device with fd %d found", fd);
+		close(fd);
+		return;
+	}
+
+	if (libseat_close_device(session->seat, dev->device_id) == -1) {
+		wlr_log_errno(WLR_ERROR, "Failed to close device %d", dev->device_id);
+	}
+
+	wl_list_remove(&dev->link);
```

Comment:
Maybe we should free `dev` at this point?

Reply:
I *guess* it would make sense to not leak things. :)